### PR TITLE
Return no trailer if Recv returns context canceled error

### PR DESCRIPTION
**Pull request by canguler**

### CODE REVIEW

File: stream.go [hunk 1]
```diff
@@ -714,6 +719,10 @@ func (cs *clientStream) RecvMsg(m interface{}) error {
 	err := cs.withRetry(func(a *csAttempt) error {
 		return a.recvMsg(m, recvInfo)
 	}, cs.commitAttemptLocked)
+	// The context is canceled.
+	if err != nil && err.Error() == toRPCErr(context.Canceled).Error() {
```
Comment (dfawley) [inline stream.go]: I'm pretty sure a server could inadvertently trick this code into suppressing the trailers from the client:

```go
func ServiceHandler(...) error {
   if err := doAThingThatMonitorsContext(ctx); err != nil {
      stream.SetTrailer(md)
      return err
   }
}
```

The proper fix is most likely in the transport layer.

File: test/context_canceled_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
```
Comment (dfawley) [inline test/context_canceled_test.go]: Happy New Year! ;)

File: test/context_canceled_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Binary wait_for_ready is an example for "wait for ready".
```
Comment (dfawley) [inline test/context_canceled_test.go]: It might be, but that's not related to this file.

File: test/context_canceled_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	"testing"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+	testpb "google.golang.org/grpc/test/grpc_testing"
+)
+
+//TestRace tests.
```
Comment (dfawley) [inline test/context_canceled_test.go]: Remove/update please.

File: test/context_canceled_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	cancel()
+	_, err = str.Recv()
+	if err == nil {
+		t.Fatalf("non-nil error expected from Recv()")
+	}
+	trl := str.Trailer()
+	if status.Code(err) == codes.PermissionDenied && trl["a"] == nil {
+		t.Fatalf("<<a>> not in trailer")
+	} else if status.Code(err) == codes.Canceled && trl["a"] != nil {
+		t.Fatalf("<<a>> in trailer")
```
Comment (dfawley) [inline test/context_canceled_test.go]: Can you do this in a loop ~100 times and make sure we get one or a few instances of both results?

File: test/context_canceled_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	str, err := ss.client.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", ss.client, err)
+	}
+	<-wait
+	time.Sleep(time.Millisecond)
```
Comment (dfawley) [inline test/context_canceled_test.go]: This is to make sure the call returns, I assume..  Is the wait channel necessary, then, given there is also a sleep?

**Comment by canguler** [comment]
Opened a new PR #2554.
Resolved ones are marked.

**Pull request closed without merging**